### PR TITLE
Make DataWord immutable

### PR DESCRIPTION
**Pull request by zilm13**
Uses #902 which is a bit outdated
+ resolve conflicts 
+ few fixes

Should resolve #902 #766 #884 #706 
@mkalinin @eugene-shevchenko

### CODE REVIEW

File: ethereumj-core/src/main/java/org/ethereum/vm/DataWord.java [hunk 1]
```diff
@@ -81,18 +86,26 @@ public DataWord(byte[] data) {
             this.data = ByteUtil.EMPTY_BYTE_ARRAY;
```
Comment (mkalinin) [inline ethereumj-core/src/main/java/org/ethereum/vm/DataWord.java]: Better use inner constant here.

File: ethereumj-core/src/main/java/org/ethereum/vm/DataWord.java [hunk 2]
```diff
@@ -81,18 +86,26 @@ public DataWord(byte[] data) {
...
         else if (data.length == 32)
             this.data = data;
-        else if (data.length <= 32)
-            System.arraycopy(data, 0, this.data, 32 - data.length, data.length);
-        else
-            throw new RuntimeException("Data word can't exceed 32 bytes: " + data);
+        else if (data.length <= 32) {
+            byte[] bytes = ZERO.getData();
```
Comment (mkalinin) [inline ethereumj-core/src/main/java/org/ethereum/vm/DataWord.java]: To preserve mutability there should be a brand new empty array allocated and then filled with given data. The same thing is with above case

File: ethereumj-core/src/main/java/org/ethereum/vm/DataWord.java [hunk 3]
```diff
@@ -81,18 +86,26 @@ public DataWord(byte[] data) {
...
+        } else
+            throw new RuntimeException(String.format("Data word can't exceed 32 bytes: 0x%s", ByteUtil.toHexString(data)));
     }
 
     public byte[] getData() {
-        return data;
+        return Arrays.copyOf(data, data.length);
+    }
+
+    public DataWord insert(int index, byte element) {
```
Comment (mkalinin) [inline ethereumj-core/src/main/java/org/ethereum/vm/DataWord.java]: This is not needed. We can use `program.stackPush(DataWord.ONE)` to cover all cases that this method covers

File: ethereumj-core/src/main/java/org/ethereum/vm/DataWord.java [hunk 4]
```diff
@@ -184,158 +197,153 @@ public boolean isNegative() {
...
         }
-        return this;
+        return new DataWord(newData);
     }
 
-    public void negate() {
+    public DataWord negate() {
 
-        if (this.isZero()) return;
+        if (this.isZero()) return zero();
```
Comment (mkalinin) [inline ethereumj-core/src/main/java/org/ethereum/vm/DataWord.java]: `return this;`

File: ethereumj-core/src/main/java/org/ethereum/vm/DataWord.java (hunk 2 above)
Comment (mkalinin) [reply ethereumj-core/src/main/java/org/ethereum/vm/DataWord.java]: But doing it that way we won't be able to efficiently use this constructor in inner `DataWord` allocations where `DataWord` itself cares about immutability. I see two options (i) make this constructor private and use for inner purposes, for outer create a method `DataWord.from(byte[] data)` with all required checks and copies (ii) do it vise versa

File: ethereumj-core/src/main/java/org/ethereum/vm/DataWord.java [hunk 5]
```diff
@@ -184,158 +197,153 @@ public boolean isNegative() {
...
     }
 
     // TODO: improve with no BigInteger
-    public void div(DataWord word) {
+    public DataWord div(DataWord word) {
 
         if (word.isZero()) {
-            this.and(ZERO);
-            return;
+            return this.and(ZERO);
```
Comment (mkalinin) [inline ethereumj-core/src/main/java/org/ethereum/vm/DataWord.java]: `return ZERO;`

**Comment by zilm13** [comment]
Second try!

File: ethereumj-core/src/main/java/org/ethereum/vm/DataWord.java (hunk 2 above)
Comment (Nashatyrev) [reply ethereumj-core/src/main/java/org/ethereum/vm/DataWord.java]: Yep, this looks pretty odd. I see that `getData()` makes a copy, but `new byte[32]` suites better here

File: ethereumj-core/src/main/java/org/ethereum/vm/DataWord.java [hunk 6]
```diff
@@ -81,18 +86,26 @@ public DataWord(byte[] data) {
...
+            System.arraycopy(data, 0, bytes, 32 - data.length, data.length);
+            this.data = bytes;
+        } else
+            throw new RuntimeException(String.format("Data word can't exceed 32 bytes: 0x%s", ByteUtil.toHexString(data)));
     }
 
     public byte[] getData() {
-        return data;
+        return Arrays.copyOf(data, data.length);
+    }
```
Comment (Nashatyrev) [inline ethereumj-core/src/main/java/org/ethereum/vm/DataWord.java]: I would leave this as is, just making sure the returned array is not accidentally modified else where (which normally shouldn't be the case for `getData()` method). If one really needs a copy I'd suggest add special `getDataCopy()` method

File: ethereumj-core/src/main/java/org/ethereum/vm/DataWord.java [hunk 7]
```diff
@@ -81,18 +86,26 @@ public DataWord(byte[] data) {
...
+
+    public DataWord insert(int index, byte element) {
+        byte[] newData = this.getData();
+        newData[index] = element;
+        return new DataWord(newData);
     }
 
     public byte[] getNoLeadZeroesData() {
-        return ByteUtil.stripLeadingZeroes(data);
+        return ByteUtil.stripLeadingZeroes(getData());
```
Comment (Nashatyrev) [inline ethereumj-core/src/main/java/org/ethereum/vm/DataWord.java]: Again `getData()` makes absolutely unnecessary copy internally

File: ethereumj-core/src/main/java/org/ethereum/vm/VM.java [hunk 8]
```diff
@@ -518,13 +518,12 @@ else if (oldValue != null && newValue.isZero()) {
                     if (logger.isInfoEnabled())
                         hint = word1.value() + " < " + word2.value();
 
+                    DataWord andResult = word1.and(DataWord.zero());
                     if (word1.value().compareTo(word2.value()) == -1) {
-                        word1.and(DataWord.ZERO);
-                        word1.getData()[31] = 1;
+                        program.stackPush(andResult.insert(31, (byte) 1));
```
Comment (Nashatyrev) [inline ethereumj-core/src/main/java/org/ethereum/vm/VM.java]: Strange code to push `1` constant

File: ethereumj-core/src/main/java/org/ethereum/vm/VM.java [hunk 9]
```diff
@@ -518,13 +518,12 @@ else if (oldValue != null && newValue.isZero()) {
...
                     } else {
-                        word1.and(DataWord.ZERO);
+                        program.stackPush(andResult);
```
Comment (Nashatyrev) [inline ethereumj-core/src/main/java/org/ethereum/vm/VM.java]: Strange code to push `0`

File: ethereumj-core/src/main/java/org/ethereum/vm/VM.java [hunk 10]
```diff
@@ -536,13 +535,12 @@ else if (oldValue != null && newValue.isZero()) {
                     if (logger.isInfoEnabled())
                         hint = word1.sValue() + " < " + word2.sValue();
 
+                    DataWord andResult = word1.and(DataWord.zero());
                     if (word1.sValue().compareTo(word2.sValue()) == -1) {
-                        word1.and(DataWord.ZERO);
-                        word1.getData()[31] = 1;
+                        program.stackPush(andResult.insert(31, (byte) 1));
```
Comment (Nashatyrev) [inline ethereumj-core/src/main/java/org/ethereum/vm/VM.java]: the same

File: ethereumj-core/src/main/java/org/ethereum/vm/VM.java [hunk 11]
```diff
@@ -536,13 +535,12 @@ else if (oldValue != null && newValue.isZero()) {
...
                     } else {
-                        word1.and(DataWord.ZERO);
+                        program.stackPush(andResult);
                     }
```
Comment (Nashatyrev) [inline ethereumj-core/src/main/java/org/ethereum/vm/VM.java]: the same

Comment (Nashatyrev) [reply ethereumj-core/src/main/java/org/ethereum/vm/VM.java]: And cases below

File: ethereumj-core/src/main/java/org/ethereum/vm/VM.java (hunk 8 above)
Comment (zilm13) [reply ethereumj-core/src/main/java/org/ethereum/vm/VM.java]: It's outdated

File: ethereumj-core/src/main/java/org/ethereum/vm/VM.java (hunk 9 above)
Comment (zilm13) [reply ethereumj-core/src/main/java/org/ethereum/vm/VM.java]: It's outdated

File: ethereumj-core/src/main/java/org/ethereum/vm/VM.java (hunk 10 above)
Comment (zilm13) [reply ethereumj-core/src/main/java/org/ethereum/vm/VM.java]: same )

File: ethereumj-core/src/main/java/org/ethereum/vm/VM.java (hunk 11 above)
Comment (zilm13) [reply ethereumj-core/src/main/java/org/ethereum/vm/VM.java]: +1

File: ethereumj-core/src/main/java/org/ethereum/vm/DataWord.java [hunk 12]
```diff
@@ -184,158 +215,153 @@ public boolean isNegative() {
         return result == 0x80;
     }
 
-    public DataWord and(DataWord w2) {
-
+    public DataWord and(DataWord word) {
+        byte[] newData = this.getData();
```
Comment (Nashatyrev) [inline ethereumj-core/src/main/java/org/ethereum/vm/DataWord.java]: It's very confusing to use `getData()` for making an array copy. It better to make it explicitly or add `copyData()` private method
The similar below.

File: ethereumj-core/src/main/java/org/ethereum/vm/DataWord.java [hunk 13]
```diff
@@ -22,77 +22,108 @@
...
+        return of(ByteBuffer.allocate(4).putInt(num).array());
+    }
+
+    public static DataWord of(long num) {
+        return of(ByteBuffer.allocate(8).putLong(num).array());
     }
 
     public byte[] getData() {
-        return data;
+        return Arrays.copyOf(data, data.length);
```
Comment (Nashatyrev) [inline ethereumj-core/src/main/java/org/ethereum/vm/DataWord.java]: I'm still suggesting returning internal array here (mentioning this in method docs). It's a normal practice (e.g. take a look at `java.nio.ByteBuffer.array()`

File: ethereumj-core/src/main/java/org/ethereum/vm/DataWord.java [hunk 14]
```diff
@@ -22,77 +22,122 @@
...
 /**
  * DataWord is the 32-byte array representation of a 256-bit number
  * Calculations can be done on this word with other DataWords
+ * DataWord is immutable. Use one of `of` factories for instance creation.
  *
  * @author Roman Mandeleil
  * @since 01.06.2014
  */
-public class DataWord implements Comparable<DataWord> {
+public final class DataWord implements Comparable<DataWord> {
```
Comment (mkalinin) [inline ethereumj-core/src/main/java/org/ethereum/vm/DataWord.java]: For some reason Github doesn't allow to comment on `public boolean isZero()` like. But this method lacks a shortcut check for whether `this == ZERO` or not

File: ethereumj-core/src/main/java/org/ethereum/vm/DataWord.java [hunk 15]
```diff
@@ -22,77 +22,122 @@
...
 
-    public DataWord(int num) {
-        this(ByteBuffer.allocate(4).putInt(num));
-    }
+    public static DataWord of(byte[] data) {
+        if (data == null || data.length == 0) {
+            return DataWord.ZERO;
+        }
+
+        boolean allExceptLastIsZero = true;
```
Comment (mkalinin) [inline ethereumj-core/src/main/java/org/ethereum/vm/DataWord.java]: You may use `8 * data.length - ByteUtil.numberOfLeadingZeros(data)`. And check the residual against `0` or `1`

File: ethereumj-core/src/main/java/org/ethereum/config/CommonConfig.java [hunk 16]
```diff
@@ -258,9 +258,9 @@ public HeaderStore headerStore() {
         return new SourceCodec<byte[], ProgramPrecompile, byte[], byte[]>(source,
                 new Serializer<byte[], byte[]>() {
                     public byte[] serialize(byte[] object) {
-                        DataWord ret = new DataWord(object);
-                        ret.add(new DataWord(1));
-                        return ret.getLast20Bytes();
+                        DataWord ret = DataWord.of(object);
+                        DataWord addResult = ret.add(DataWord.of(1));
```
Comment (mkalinin) [inline ethereumj-core/src/main/java/org/ethereum/config/CommonConfig.java]: Why not use `DataWord.ONE` here?

File: ethereumj-core/src/main/java/org/ethereum/vm/program/Program.java [hunk 17]
```diff
@@ -222,15 +222,15 @@ public byte getPreviouslyExecutedOp() {
     }
 
     public void stackPush(byte[] data) {
-        stackPush(new DataWord(data));
+        stackPush(DataWord.of(data));
     }
 
     public void stackPushZero() {
-        stackPush(new DataWord(0));
+        stackPush(DataWord.of(0));
```
Comment (mkalinin) [inline ethereumj-core/src/main/java/org/ethereum/vm/program/Program.java]: `DataWord.ZERO`

File: ethereumj-core/src/main/java/org/ethereum/vm/program/Program.java [hunk 18]
```diff
@@ -222,15 +222,15 @@ public byte getPreviouslyExecutedOp() {
...
     }
 
     public void stackPushOne() {
-        DataWord stackWord = new DataWord(1);
+        DataWord stackWord = DataWord.of(1);
```
Comment (mkalinin) [inline ethereumj-core/src/main/java/org/ethereum/vm/program/Program.java]: `DataWord.ONE`

File: ethereumj-core/src/test/java/org/ethereum/datasource/RepoNewTest.java [hunk 19]
```diff
@@ -155,31 +155,31 @@ public void testStorage1() throws Exception {
         RepositoryRoot repo = new RepositoryRoot(stateDb, null);
         byte[] addr1 = decode("aaaa");
         repo.createAccount(addr1);
-        repo.addStorageRow(addr1, new DataWord(1), new DataWord(111));
+        repo.addStorageRow(addr1, DataWord.of(1), DataWord.of(111));
```
Comment (mkalinin) [inline ethereumj-core/src/test/java/org/ethereum/datasource/RepoNewTest.java]: `DataWord.ONE`

Comment (Nashatyrev) [reply ethereumj-core/src/test/java/org/ethereum/datasource/RepoNewTest.java]: @mkalinin Here you probably a little bit doebalsa :) This is just a test

Comment (mkalinin) [reply ethereumj-core/src/test/java/org/ethereum/datasource/RepoNewTest.java]: Ahahaha, maybe just a little bit

File: ethereumj-core/src/main/java/org/ethereum/vm/DataWord.java [hunk 20]
```diff
@@ -22,77 +22,118 @@
...
-        if (data == null)
-            this.data = ByteUtil.EMPTY_BYTE_ARRAY;
-        else if (data.length == 32)
-            this.data = data;
-        else if (data.length <= 32)
-            System.arraycopy(data, 0, this.data, 32 - data.length, data.length);
-        else
-            throw new RuntimeException("Data word can't exceed 32 bytes: " + data);
+    public static DataWord of(long num) {
+        return of(ByteBuffer.allocate(8).putLong(num).array());
```
Comment (eugene-shevchenko) [inline ethereumj-core/src/main/java/org/ethereum/vm/DataWord.java]: same here `Long.BYTES`

File: ethereumj-core/src/main/java/org/ethereum/vm/DataWord.java [hunk 21]
```diff
@@ -22,77 +22,118 @@
...
+        byte[] bb = new byte[32];
+        bb[31] = num;
+        return new DataWord(bb);
+
     }
 
-    public DataWord(ByteArrayWrapper wrappedData){
-        this(wrappedData.getData());
+    public static DataWord of(int num) {
+        return of(ByteBuffer.allocate(4).putInt(num).array());
```
Comment (eugene-shevchenko) [inline ethereumj-core/src/main/java/org/ethereum/vm/DataWord.java]: It's better use `Integer.BYTES` constant.

File: ethereumj-core/src/main/java/org/ethereum/vm/DataWord.java [hunk 22]
```diff
@@ -22,77 +22,118 @@
...
 
+    /**
+     * Returns instance data
+     * Actually copy of internal byte array is provided
+     * in order to protect DataWord immutability
+     * @return instance data
+     */
     public byte[] getData() {
-        return data;
+        return Arrays.copyOf(data, data.length);
```
Comment (eugene-shevchenko) [inline ethereumj-core/src/main/java/org/ethereum/vm/DataWord.java]: Maybe it's better to use `copyData()` and mark `getData()` as deprecated.

File: ethereumj-core/src/main/java/org/ethereum/vm/DataWord.java (hunk 21 above)
Comment (Nashatyrev) [reply ethereumj-core/src/main/java/org/ethereum/vm/DataWord.java]: We've got this already: `org.ethereum.util.ByteUtil.longToBytes`

File: ethereumj-core/src/main/java/org/ethereum/vm/DataWord.java (hunk 20 above)
Comment (Nashatyrev) [reply ethereumj-core/src/main/java/org/ethereum/vm/DataWord.java]: We've got this already: `org.ethereum.util.ByteUtil.longToBytes`

**Pull request merged**